Implemented nuget.exe add and nuget.exe init commands

deepakaravindr opened the pull request:
Both nuget.exe add and nuget.exe init follows the new v3 model to only copy the nupkg, extract the nuspec and create a SHA512 of the nupkg file at the destination. With a provision to expand the packages using '-Expand' switch

nuget.exe add: https://github.com/NuGet/Home/issues/1421
Adds a given package to the specified source if it is a file source. http sources are not supported.

nuget.exe init: https://github.com/NuGet/Home/issues/1423
Adds all the packages from a given feed to the specified destination feed if it is a file source feed. http feeds are not supported.

Following are pending and/or not covered in this change.
1. nuget.exe init only takes a folder of nupkgs as the source. Should this be upgraded to support every v2 folder source and probably v3 folder sources as well? 
2. nuget.exe init could parallelize package additions.
3. Need more tests for the new '-Expand' switch. Also, need to verify and update the messages for the -Expand switch as needed.
4. Marker file for nuget.exe add and init: https://github.com/NuGet/Home/issues/1456
5. Concurrency for "multiple exes" on other platforms: https://github.com/NuGet/Home/issues/1455
6. With the latest design, Updating nuget clients may not be necessary anymore. We should close https://github.com/NuGet/Home/issues/1425 and https://github.com/NuGet/Home/issues/1424 as Won't Fix.
7. Update documentation on docs.nuget.org

@yishaigalatzer @MeniZalzman @emgarten

deepakaravindr committed:
Updated NuGetPackageUtils to be able to extract only the nuspec file. This will be used by nuget.exe add and nuget.exe init

deepakaravindr committed:
nuget.exe init and nuget.exe add with tests, that supports a
user-defined offline feed as controlled by settings.

deepakaravindr committed:
nuget.exe add and nuget.exe init which are simpler and always needs a
mandatory source and destination feed for add and init respectively.
Need to fix tests

deepakaravindr committed:
Fixed the tests and the product to make source switch for nuget.exe add
and destination feed for nuget.exe init mandatory.

deepakaravindr committed:
Added support for expanding the packages in case of nuget.exe add
and nuget.exe init using an -Expand switch. Need to add tests

deepakaravindr committed:
Fixed maxArgs for nuget.exe add and added tests for the same

deepakaravindr committed:
1) Made nuget.exe init create the destination directory if it is not found.
2) Added more tests for nuget.exe init

yishaigalatzer commented:
1. It should take any existing folder source
2. Lets skip the parallelization for now
3. Add the tests
4. Skip NuGet/Home#1456 
5. Concurrency is important, but can be done in a follow up PR
6. Closed the items
7. Create a doc PR

deepakaravindr committed:
Updated init to support both v2 style and v3 style folder sources.
Verified manually that it works. Need to add tests

deepakaravindr committed:
Added tests for v2-style and v3-style source folders.

deepakaravindr committed:
Added tests for -Expand switch for nuget.exe add and init

deepakaravindr commented:
- It takes any existing folder source. V2 and V3. Added tests. Manually tested.
- OK
- Added tests for -Expand switch. Manually tested
- OK
- OK
- OK
- Will do

emgarten commented on src/NuGet.Clients/NuGet.CommandLine/NuGetCommand.resx:
```diff
@@ -5245,4 +5245,41 @@ nuget update -Self</value>
     <value>Specifies the version of MSBuild to be used with this command. Supported values are 4, 12, 14. By default the MSBuild in your path is picked, otherwise it defaults to the highest installed version of MSBuild.</value>
   </data>
+  <data name="AddCommandDescription" xml:space="preserve">
+    <value>Adds a given package to the specified source if it is a file source. http sources are not supported.</value>
```
This doesn't really describe what the command is for

MeniZalzman replied:
+1, we need to explain the offline feed concept here

deepakaravindr replied:
Discussed with @yishaigalatzer and updated the command description both for add and init. Check it out in the next iteration.

emgarten commented:
:shipit:

MeniZalzman commented on src/NuGet.Clients/NuGet.CommandLine/Commands/AddCommand.cs:
```diff
@@ -0,0 +1,54 @@
+    public class AddCommand : Command
+    {
+        [Option(typeof(NuGetCommand), "AddCommandSourceDescription", AltName = "src")]
+        public string Source { get; set; }
```
Please check if we have a way to declare mandatory parameters

deepakaravindr replied:
No. We do not have a way today to declare mandatory parameters.

MeniZalzman commented on src/NuGet.Clients/NuGet.CommandLine/Commands/AddCommand.cs:
```diff
@@ -0,0 +1,54 @@
+            // Arguments[0] will not be null at this point.
+            // Because, this command has MinArgs set to 1.
+            var packagePath = Arguments[0];
+            OfflineFeedUtility.ValidatePath(packagePath);
```
As we discussed let change the Validate method to do all the needed validations and also change the method name to ThrowIfInvalid~

deepakaravindr replied:
Renamed to ThrowIfInvalid and added another one which is ThrowIfInvalidOrNotFound and have used the methods as appropriate.

MeniZalzman commented on src/NuGet.Clients/NuGet.CommandLine/Commands/AddCommand.cs:
```diff
@@ -0,0 +1,54 @@
+                throwIfPackageExists: false,
+                expand: Expand);
+
+            await OfflineFeedUtility.AddPackageToSource(offlineFeedAddContext, CancellationToken.None);
```
Can this throw? if so which exception the user will get?

deepakaravindr replied:
OfflineFeedAddContext, which is passed in to the method, has that information. It will always throw CommandLineException

MeniZalzman commented on src/NuGet.Clients/NuGet.CommandLine/Commands/InitCommand.cs:
```diff
@@ -0,0 +1,155 @@
+            var destination = Arguments[1];
+            OfflineFeedUtility.ValidatePath(source);
+
+            if (!Directory.Exists(source))
```
Same here

deepakaravindr committed:
1) Added a new method ThrowIfInvalidOrNotFound to OfflineFeedUtility to
simplify the not exists check on the path both for add and init commands.

2) Renamed ValidatePath to ThrowIfInvalid.

3) Added tests for OfflineFeedUtility

deepakaravindr committed:
Incorporated the comments on the help string messages.

MeniZalzman commented:
:shipit:

deepakaravindr committed:
Removed NuGetPackageUtilsTests and
moved the tests to NugetPackageUtilTests.cs.
Also, fixed the existing tests to use the new signature for
InstallFromSourceAsync

deepakaravindr commented:
Merged with 

https://github.com/NuGet/NuGet.Client/commit/30046491fb164a94a38a36f7b384b6fb2904d83c,

https://github.com/NuGet/NuGet.Client/commit/5eefae084d91f62b2fe0a10559f15d9fd39f5639
and
https://github.com/NuGet/NuGet.Client/commit/5272f060f2162c93b8d26f16fb365c5beed8d23e
